src/Internal/Service/OrderService.php

```diff
@@ -48,4 +85,45 @@ public function __construct( WC_Payments_Order_Service $legacy_service ) {
 	public function set_payment_method_id( int $order_id, string $payment_method_id ) {
 		$this->legacy_service->set_payment_method_id_for_order( $order_id, $payment_method_id );
 	}
+
+	/**
+	 * Generates payment metadata from order details.
+	 *
+	 * @param int          $order_id     ID of the order.
+	 * @param Payment_Type $payment_type Type of the payment (recurring or not).
+	 * @return array                     The metadat athat will be sent to the server.
+	 * @throws Order_Not_Found_Exception
+	 */
+	public function get_payment_metadata( int $order_id, Payment_Type $payment_type = null ) {
+		$order = $this->get_order( $order_id );
+
+		$name     = sanitize_text_field( $order->get_billing_first_name() ) . ' ' . sanitize_text_field( $order->get_billing_last_name() );
+		$email    = sanitize_email( $order->get_billing_email() );
+		$metadata = [
+			'customer_name'        => $name,
+			'customer_email'       => $email,
+			'site_url'             => esc_url( get_site_url() ),
+			'order_id'             => $order->get_id(),
+			'order_number'         => $order->get_order_number(),
+			'order_key'            => $order->get_order_key(),
+			'payment_type'         => $payment_type,
+			'checkout_type'        => $order->get_created_via(),
+			'client_version'       => WCPAY_VERSION_NUMBER,
+			'subscription_payment' => 'no',
+		];
+
+		if (
+			'recurring' === (string) $payment_type
+			&& $this->legacy_proxy->function_exists( 'wcs_order_contains_subscription' )
+			&& $this->legacy_proxy->call_function( 'wcs_order_contains_subscription', $order, 'any' )
+		) {
+			$use_stripe_billing = $this->legacy_proxy->call_static( WC_Payments_Features::class, 'should_use_stripe_billing' );
+			$is_renewal         = $this->legacy_proxy->call_function( 'wcs_order_contains_renewal', $order );
+
+			$metadata['subscription_payment'] = $is_renewal ? 'renewal' : 'initial';
+			$metadata['payment_context']      = $use_stripe_billing ? 'wcpay_subscription' : 'regular_subscription';
+		}
+
+		return apply_filters( 'wcpay_metadata_from_order', $metadata, $order, $payment_type );
```

Comment:
I don't think we've discussed such plans, though in this case, I think this is a very meaningful filter, and it makes sense to keep it.

I wouldn't mind opening another issue, in case you have anything specific in mind. For this instance, and especially for now, we should avoid breaking changes, caused by the payment process.

Reply:
By the way, we have the `HooksProxy`, and at least now the call is unit-tested: https://github.com/Automattic/woocommerce-payments/pull/7333/commits/de977fef419d9d29a7ca3e787f2b742e26e4f13f